firmware/src/core/utils/HTTPClientWrapper.h

```diff
@@ -0,0 +1,57 @@
+#ifndef HTTP_CLIENT_WRAPPER_H
+#define HTTP_CLIENT_WRAPPER_H
+
+#include <Arduino.h>
+#include <HTTPClient.h>
+#include <freertos/FreeRTOS.h>
+#include <freertos/semphr.h>
+#include <freertos/task.h>
+#include <freertos/queue.h>
+#include <functional>
+
+class HTTPClientWrapper {
+public:
+    using ResponseCallback = std::function<void(int httpCode, const String& response)>;
+    
+    static HTTPClientWrapper* getInstance();
+    bool makeRequest(const String& url, ResponseCallback callback);
+    bool addRequest(const String& url, ResponseCallback callback);
+    void processRequestQueue();  
+    void processResponseQueue();
+
+private:
+    // Add debug counters
+    static volatile uint32_t activeRequests;
+    static volatile uint32_t maxConcurrentRequests;
+
+    HTTPClientWrapper();
+    ~HTTPClientWrapper();
+    
+    static void httpTask(void* params);
+    
+    struct RequestParams {
+        String url;
+        ResponseCallback callback;
+    };
+
+    struct ResponseData {
+        int httpCode;
+        String response;
+        ResponseCallback callback;
+    };
+
+    static HTTPClientWrapper* instance;
+    static SemaphoreHandle_t requestSemaphore;
+    static QueueHandle_t requestQueue;
+    static QueueHandle_t responseQueue;
```

Comment:
There is no point in the responseQueue?
There should always be 1 or 0 responses, so either use a queue size of 1, or remove it entirely?

Reply:
I disagree.  The request is spun off on a Task thread that is independent of the main loop.  Yes only one can run at a time but since we're processing the callback from the main loop in theory it could be possible to get one response stepping on another.  If we DON'T process the callback in the main loop then we could eliminate the responseQueue but that also means your processing response data and updating things OUTSIDE the main loop that could cause race conditions.   The responseQueue limits that risk as it allows us to keep widget processing in the main loop.